src/api_gen.jl

```diff
@@ -21,14 +28,65 @@ const api_methods = [f[:name] => f for f in _functions]
 const typemap = @compat Dict{Symbol,Type}(
     :Integer => Integer,
     :Boolean => Bool,
-    :String => Union(ByteString, Vector{Uint8}),
+    :String => Union(ByteString, Vector{Uint8})
 )
 
-# Types defined by the api
-immutable NvimApiObject{N} <: NvimObject
-    client::NvimClient
-    # TODO: use a fixarray or Uint64
-    hnd::Vector{Uint8}
+function parsetype(t::Symbol)
+    components = split(rstrip(string(t), ')'), '(')
+    top = symbol(components[1])
+    params = length(components) > 1 ? split(components[2], ", ") : []
+    (top, params)
+end
+
+function gettype(t::Symbol)
+    top, params = parsetype(t)
+    if haskey(typemap, top)
+        typemap[top]
+    elseif top == :ArrayOf
+        tcontains = gettype(symbol(params[1]))
+        if length(params) == 1
+            if isa(tcontains, UnionType)
+                tarr = Union([Array{t, 1} for t in tcontains.types]...)
+            else
+                tarr = Array{tcontains, 1}
+            end
+            Union(tarr, Array{None, 1})
+        else
+            @compat Tuple{fill(tcontains, int(params[2]))...}
+        end
+    else
+        Any
+    end
+end
+
+convertdict(val::Dict) = Dict{ByteString,Any}([(convertdict(k),convertdict(v)) for (k,v) in val])
+convertdict(val::Union(ByteString, Vector{Uint8})) = bytestring(val)
+convertdict(val::Vector) = [convertdict(v) for v in val]
+convertdict(val) = val
+
+function retconvert(c::NvimClient, t::Symbol, v=:void)
+    ttop, tparams = parsetype(t)
+
+    if ttop == :ArrayOf
+        tcontains = symbol(tparams[1])
+        if length(tparams) == 1
+            map(x->retconvert(c, tcontains, x), v)
+        else
+            ntuple(i->retconvert(c, tcontains, v[i]), int(tparams[2]))
+        end
+    elseif ttop == :Array || ttop == :Boolean || ttop == :Integer
```

Comment:
Also in the (untyped) Array and Object case we need to do the recursive conversions (`convertdict`).  I suppose, in the dispatch approach, what you call `convertdict` could be the `retconvert(::Type{Any}, ...)` fallback case.